### ISSUE TITLE
Kernel 5.15

### ISSUE THREAD

**Issue by builderjer** [viraniac/aiy-voicebonnet-soundcard-dkms#1]
I am trying to get this to install with kernel `5.15.76-v8+` but keep getting a build error

```
cat /var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build/make.log 
DKMS make.log for aiy-voicebonnet-soundcard-3.0 for kernel 5.15.76-v8+ (aarch64)
Thu  8 Dec 19:27:28 GMT 2022
make: Entering directory '/usr/src/linux-headers-5.15.76-v8+'
  CC [M]  /var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build/snd-aiy-voicebonnet.o
  CC [M]  /var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build/rt5645.o
  CC [M]  /var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build/rl6231.o
In file included from /var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build/snd-aiy-voicebonnet.c:25:
/var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build/rt5645.h:12:10: fatal error: sound/rt5645.h: No such file or directory
   12 | #include <sound/rt5645.h>
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [scripts/Makefile.build:289: /var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build/snd-aiy-voicebonnet.o] Error 1
make[1]: *** Waiting for unfinished jobs....
In file included from /var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build/rt5645.c:32:
/var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build/rt5645.h:12:10: fatal error: sound/rt5645.h: No such file or directory
   12 | #include <sound/rt5645.h>
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [scripts/Makefile.build:289: /var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build/rt5645.o] Error 1
make: *** [Makefile:1900: /var/lib/dkms/aiy-voicebonnet-soundcard/3.0/build] Error 2
make: Leaving directory '/usr/src/linux-headers-5.15.76-v8+'
```

Any help would be great as I have 3 of these that I want to put to use

**Comment by viraniac** [comment]
Could you please try the files from [here](https://github.com/viraniac/aiy-voicebonnet-soundcard-dkms/releases/tag/3.0-1.1-beta)

PS: I am not a kernel developer either. So these are mostly just on best effort basis. You might have to play around with various settings in alsa mixer to get clear audio.

To disable the builtin sound card, you might want to disable the hdmi audio by modifying the following value in /boot/config.txt file

Before
`dtoverlay=vc4-kms-v3d`

After
`dtoverlay=vc4-kms-v3d,audio=off`

**Comment by viraniac** [comment]
If its not required for you to use debian bullseye based Raspberry pi OS, I will suggest to use Raspberry Pi OS Lite (Legacy) version instead

**Comment by builderjer** [comment]
What I am actually trying to use it on is Manjaro with a dkms kernel.  It is for a DIY smart speaker using [OVOS](https://github.com/OpenVoiceOS/ovos-image-arch-recipe)

The image works well, but I have a few of these that I want to use.  
I will give it a go with all OS's that I use

**Comment by viraniac** [comment]
Just out of curiosity, I noticed that your kernel version ends with v8+, are you using the hat on a non-zero pi or are you trying to get it working on zero 2w? I have been trying to get a zero 2w for myself, but it never seems to be in stock in India.

**Comment by builderjer** [comment]
It's a Pi4 8G.  But yes, I have a couple of Pi0's id like to try it on also

**Comment by builderjer** [comment]
it installed, do I need to add an overlay also?

**Comment by viraniac** [comment]
Overlay gets applied automatically on voicebonnet (voice kit v2). if its successful, you should see the card listed in the output of `aplay -l`

**Comment by builderjer** [comment]
```
aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: vc4hdmi1 [vc4-hdmi-1], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```
but I don't seem to be getting any mic input.  But, it IS installed and seems to work.  I will continue messing with this to get it to work.  Thank you VERY MUCH

**Comment by viraniac** [comment]
That is not the output you were supposed to get. Your output should be similar to this

```
pi@raspberrypi:~ $ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: aiyvoicebonnet [aiy-voicebonnet], device 0: Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0 [Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0

pi@raspberrypi:~ $ arecord -l
**** List of CAPTURE Hardware Devices ****
card 0: aiyvoicebonnet [aiy-voicebonnet], device 0: Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0 [Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0

pi@raspberrypi:~ $ dpkg -l | grep aiy
ii  aiy-dkms                             2.0-1.1                          all          Hardware device drivers for AIY Projects Kits.
ii  aiy-voicebonnet-soundcard-dkms       3.0-1.1                          all          AIY Voice DKMS driver.

pi@raspberrypi:~ $ grep audio /boot/config.txt
# uncomment to force a HDMI mode rather than DVI. This can make audio work in
# Enable audio (loads snd_bcm2835)
#dtparam=audio=on
dtoverlay=vc4-kms-v3d,audio=off

```

Also you will have to reboot the pi after installing those two deb files

**Comment by builderjer** [comment]
I did not install `aiy-dkms`. Do I get it from the original Google repo?

**Comment by viraniac** [comment]
No, the aiy-dkms package from google repo fails to install as well. You can find the updated package in the same link that I uploaded earlier

**Comment by builderjer** [comment]
sorry, I missed that on my first pass.  I'm actually burning the image that I want to use this on and will continue testing from there.

**Comment by builderjer** [comment]
do you have the source for `aiy-dkms` instead of a .deb file?
is [this it](https://github.com/viraniac/aiyprojects-raspbian/tree/aiyprojects/drivers/aiy/debian)

**Comment by viraniac** [comment]
yes, its here - [https://github.com/viraniac/aiyprojects-raspbian/tree/aiyprojects/drivers/aiy](https://github.com/viraniac/aiyprojects-raspbian/tree/aiyprojects/drivers/aiy)

**Comment by builderjer** [comment]
So i made it to this
```
 dkms status
aiy-dkms/2.1.1, 5.15.81-1-MANJARO-ARM-RPI, aarch64: installed (original_module exists)
snd-aiy-voicebonnet/3.1.1b, 5.15.81-1-MANJARO-ARM-RPI, aarch64: installed
```

but `aplay -l`
 does not show any devices

**Comment by viraniac** [comment]
I just tried it on Manjaro's minimal image and seems to get detected fine for me.

```
[pi@raspberrypi ~]$ uname -a
Linux raspberrypi 5.15.78-1-MANJARO-ARM-RPI #1 SMP PREEMPT Sat Nov 12 17:13:38 UTC 2022 aarch64 GNU/Linux

[pi@raspberrypi ~]$ dkms status
aiy-dkms/2.0-1.1, 5.15.78-1-MANJARO-ARM-RPI, aarch64: installed
aiy-voicebonnet-soundcard/3.0-1.1, 5.15.78-1-MANJARO-ARM-RPI, aarch64: installed

[pi@raspberrypi ~]$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: Headphones [bcm2835 Headphones], device 0: bcm2835 Headphones [bcm2835 Headphones]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7
card 1: vc4hdmi [vc4-hdmi], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 2: aiyvoicebonnet [aiy-voicebonnet], device 0: Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0 [Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0

[pi@raspberrypi ~]$ arecord -l
**** List of CAPTURE Hardware Devices ****
card 2: aiyvoicebonnet [aiy-voicebonnet], device 0: Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0 [Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

**Comment by viraniac** [comment]
Are you sure you have voice bonnet aka [voice-kit v2](https://aiyprojects.withgoogle.com/static/images/voice-v2/guide/voice-003-1.jpg) and not the [voice-kit v1](https://aiyprojects.withgoogle.com/static/images/voice-v1/materials.jpg)?

**Comment by builderjer** [comment]
OK, raspiOS 64 bit because I am running on an 8G pi
This is my output from the above commands.
```
aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: vc4hdmi1 [vc4-hdmi-1], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0

arecord -l
**** List of CAPTURE Hardware Devices ****

dpkg -l | grep aiy
ii  aiy-dkms                                      2.0-1.1                           all          Hardware device drivers for AIY Projects Kits.
ii  aiy-voicebonnet-soundcard-dkms                3.0-1.1                           all          AIY Voice DKMS driver.

cat /boot/config.txt 
# uncomment to force a HDMI mode rather than DVI. This can make audio work in
# DMT (computer monitor) modes
hdmi_drive=2

# Uncomment some or all of these to enable the optional hardware interfaces
dtparam=i2c_arm=on
dtparam=i2s=on
dtparam=spi=on

# Enable audio (loads snd_bcm2835)
#dtparam=audio=on

# Enable DRM VC4 V3D driver
dtoverlay=vc4-kms-v3d,audio=off

# Run in 64-bit mode
arm_64bit=1
```
I would really love to get this working

**Comment by builderjer** [comment]
also getting this in `dmesg` not sure that it is relevant though
```
[   30.327792] hdmi-audio-codec hdmi-audio-codec.0.auto: ASoC: error at snd_soc_dai_startup on i2s-hifi: -19
[   30.327804]  MAI: soc_pcm_open() failed (-19)
[   33.759777] cam-dummy-reg: disabling
```

**Comment by builderjer** [comment]
> Are you sure you have voice bonnet aka [voice-kit v2](https://aiyprojects.withgoogle.com/static/images/voice-v2/guide/voice-003-1.jpg) and not the [voice-kit v1](https://aiyprojects.withgoogle.com/static/images/voice-v1/materials.jpg)?

yes, I have another also, maybe the card is bad itself

**Comment by builderjer** [comment]
Ok, found my problem. I had a POE hat under the voicebonnet that was preventing it from being detected. But now it seems it is not capturing audio. I will mess with some mixer settings and see what happens

**Comment by builderjer** [comment]
When this is up and working for me, and I'm close, just didn't have time today, I would like to use this as a plugin for [ovos](openvoiceos.com). Can you add a license, so I can use your code and the ovos dev team will be satisfied also? I know that the [Google version](https://github.com/google/aiyprojects-raspbian) is licensed  with apache 2.0 and i think that should cary on unless you choose to licence differently. That would be awesome.

**Comment by viraniac** [comment]
Consider it to be Apache 2.0 licensed. Same as before. License is mentioned [here](https://github.com/viraniac/aiy-voicebonnet-soundcard-dkms/blob/master/debian/copyright) and [here](https://github.com/viraniac/aiyprojects-raspbian/blob/aiyprojects/LICENSE)

Once you are done with your changes, please share the same. That will help others that might try to use the deb files as well.

**Comment by tcsabina** [comment]
I just did an update on my pi zero 2w, going to bullseye. This is an original Google AIY Voice Kit 2 (with Voice Bonnet).
And I can confirm that these .deb files are working fine @viraniac . Thanks for your work and effort!

```
pi@raspberrypi:~ $ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: aiyvoicebonnet [aiy-voicebonnet], device 0: Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0 [Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
pi@raspberrypi:~ $

```

```
pi@raspberrypi:~ $ arecord -l
**** List of CAPTURE Hardware Devices ****
card 0: aiyvoicebonnet [aiy-voicebonnet], device 0: Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0 [Google AIY Voice Bonnet SoundCard HiFi rt5645-aif1-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
pi@raspberrypi:~ $

```

```
pi@raspberrypi:~ $ uname -a
Linux raspberrypi 5.15.84+ #1613 Thu Jan 5 11:58:09 GMT 2023 armv6l GNU/Linux
pi@raspberrypi:~ $

```

```
pi@raspberrypi:~ $ dpkg -l | grep aiy
ii  aiy-dkms                               2.0-1.1                          all          Hardware device drivers for AIY Projects Kits.
ii  aiy-python-wheels                      1.4-1                            all          Prebuilt Python wheels for AIY
ii  aiy-voicebonnet-soundcard-dkms         3.0-1.1                          all          AIY Voice DKMS driver.
pi@raspberrypi:~ $

```

So I think this issue can be closed, as this is working on 5.15.

**Comment by viraniac** [comment]
closing as requested
